### PR TITLE
crypto/cloudflare/bn256: fix in-place addition and unmarshalling

### LINKED ISSUE #23346: bn256.G2 points fail on in-place addition
Note that this is not a security issue, since `bn256.G1.Add` is only used in `vm.runBn256Add`, and is not used in-place, there. Also, a [fix has been provided upstream](https://github.com/cloudflare/bn256/pull/12), and just hasn't been pulled in. My recommendation to fix this would be to make [these changes](https://github.com/cloudflare/bn256/pull/12/files) to to `go-ethereum`'s `cloudflare/bn256`.

A failing test is provided below, in the "Expected Behavior" section.

#### System information

Geth `version:` `geth version` => `1.15` (master) 
OS & Version: Windows/Linux/OSX
```
uname -a
Linux ubuntu-VirtualBox 5.4.0-74-generic #83-Ubuntu SMP Sat May 8 02:35:39 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

Commit hash : (if `develop`) `28ba686cb`

#### Expected behaviour

Following test of `bn256.G1` addition should pass:

```golang
func TestG2SelfAddition(t *testing.T) {
	s, _ := rand.Int(rand.Reader, Order)
	p := new(G2).ScalarBaseMult(s)
	if _, err := p.Unmarshal(p.Add(p, p).Marshal()); err != nil {
		t.Fatalf("p.Add(p, p) ∉ G₂: %v", err)
	}
}
```

#### Actual behaviour

```
ubuntu@ubuntu-VirtualBox:~/alien/go-ethereum/crypto/bn256/cloudflare$ go test -run  TestG2SelfAddition
--- FAIL: TestG2SelfAddition (0.00s)
    bn256_test.go:122: p.Add(p, p) ∉ G₂: bn256: coordinate exceeds modulus
FAIL
exit status 1
FAIL	github.com/ethereum/go-ethereum/crypto/bn256/cloudflare	0.005s
```

#### Steps to reproduce the behaviour

- Add the above test to `crypto/bn256/cloudflare/bn256_test.go` on `master`
- `go test -run  TestG2SelfAddition`

#### Backtrace

N/A

### PR DESCRIPTION
**Pull request by gballet**
Fixes #23346 

There were two upstream fixes required: one is for in-place addition and the second one was for in-place [unmarshalling](https://github.com/cloudflare/bn256/commit/d078d2bbd8e91bc6fc4eb466a2b74fcf785a938e)

### CODE REVIEW

File: crypto/bn256/cloudflare/bn256_test.go [hunk 1]
```diff
@@ -92,6 +92,19 @@ func TestTripartiteDiffieHellman(t *testing.T) {
...
+func TestG2SelfAddition(t *testing.T) {
+	s, _ := rand.Int(rand.Reader, Order)
+	p := new(G2).ScalarBaseMult(s)
+
+	if !p.p.IsOnCurve() {
+		t.Fatal("p isn't on curve")
+	}
+	m := p.Add(p, p).Marshal()
+	if _, err := p.Unmarshal(m); err != nil {
+		t.Fatalf("p.Add(p, p) ∉ G₂: %v", err)
```
Comment (MariusVanDerWijden) [inline crypto/bn256/cloudflare/bn256_test.go:104]: That error message is a bit crazy imo

**Comment by holiman** [comment]
For reference: https://github.com/cloudflare/bn256/pull/12/files

**Review by holiman — APPROVED** [review]
LGTM

**Comment by holiman** [comment]
@gballet for some reason I can't find this PR in the travis build history, can you maybe rebase it to trigger travis again?

**Pull request merged**